File: Leetcode/Arrays/Medium/55_jump_game.py
# Runtime: 84 ms, faster than 35.55% of Python online submissions for Jump Game.
# Memory Usage: 13.5 MB, less than 10.76% of Python online submissions for Jump Game.

class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        
        size = len(nums)
        if size == 1:
            return True
        
        i = size - 1
        f = False
        
        while i > 0:
            i -= 1
            d = 1
            f = False
            
            while i >= 0:
                if nums[i] >=d:
                    f = True
                    break
                else:
                    d += 1
                    i -= 1
            
            if not f:
                break
        
        if not f:
            return False
        else :
            return True

File: Leetcode/Arrays/Medium/55_jump_game.cpp
// Runtime: 8 ms, faster than 98.21% of C++ online submissions for Jump Game.
// Memory Usage: 9.7 MB, less than 90.43% of C++ online submissions for Jump Game.

class Solution {
public:
    bool canJump(vector<int>& nums) {
        
        int size = nums.size();
        int curr = size - 1;
        
        for(int i = size - 1; i >=0 ; i--) {
            if (nums[i] + i >= curr) {
                curr = i;
            }
        }
        
        if(curr == 0) {
            return true;
            
        return false;
    }
};